src/visitors/semantic/completion/mod.rs

```diff
@@ -0,0 +1,152 @@
+use std::cell::RefCell;
+use std::rc::Rc;
+
+use crate::protocol::properties::Position;
+use crate::protocol::responses::{
+    CompletionItem, CompletionItemKind, InsertTextFormat,
+};
+use crate::stdlib::Completable;
+use crate::visitors::semantic::utils;
+
+use flux::semantic::nodes::*;
+use flux::semantic::types::MonoType;
+use flux::semantic::walk::{Node, Visitor};
+
+#[derive(Default)]
+pub struct CompletableFinderState {
+    pub completables: Vec<Rc<dyn Completable>>,
+}
+
+pub struct CompletableFinderVisitor {
+    pub pos: Position,
+    pub state: Rc<RefCell<CompletableFinderState>>,
+}
+
+impl CompletableFinderVisitor {
+    pub fn new(pos: Position) -> Self {
+        CompletableFinderVisitor {
+            state: Rc::new(RefCell::new(
+                CompletableFinderState::default(),
+            )),
+            pos,
+        }
+    }
+}
+
+fn get_var_type(expr: Expression) -> Option<VarType> {
```

Comment:
you can also borrow here

`fn get_var_type(expr: &Expression) -> Option<VarType> {`

Reply:
Ownership with these visitors using walk in libflux can get tricky, prefer to clone than untangle ownership